x-pack/auditbeat/tracing/tracefs.go

```diff
@@ -0,0 +1,356 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build linux
+
+package tracing
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/joeshaw/multierror"
+)
+
+const (
+	debugFSTracingPath = "/sys/kernel/debug/tracing"
+	traceFSPath        = "/sys/kernel/tracing"
+)
+
+var (
+	kprobeRegexp *regexp.Regexp
+	formatRegexp *regexp.Regexp
+)
+
+// TraceFS is an accessor to manage event tracing via tracefs or debugfs.
+type TraceFS struct {
+	basePath string
+}
+
+func init() {
+	var err error
+	kprobeRegexp, err = regexp.Compile("^([pr]):(?:([^/ ]*)/)?([^/ ]+) ([^ ]+) ?(.*)")
+	if err != nil {
+		panic(err)
+	}
+
+	formatRegexp, err = regexp.Compile("\\s+([^:]+):([^;]*);")
+	if err != nil {
+		panic(err)
+	}
+}
+
+// NewTraceFS creates a new accessor for the event tracing feature.
+// It autodetects a tracefs mounted on /sys/kernel/tracing or via
+// debugfs at /sys/kernel/debug/tracing.
+func NewTraceFS() (*TraceFS, error) {
+	var errs multierror.Errors
+	ptr, err := NewTraceFSWithPath(traceFSPath)
+	if err != nil {
+		errs = append(errs, err)
+		ptr, err = NewTraceFSWithPath(debugFSTracingPath)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			errs = nil
+		}
+	}
+	return ptr, errs.Err()
+}
+
+// NewTraceFSWithPath creates a new accessor for the event tracing feature
+// at the given path.
+func NewTraceFSWithPath(path string) (*TraceFS, error) {
+	if _, err := os.Stat(filepath.Join(path, kprobeCfgFile)); err != nil {
+		return nil, err
+	}
+	return &TraceFS{basePath: path}, nil
+}
+
+// ListKProbes lists the currently installed kprobes / kretprobes
+func (dfs *TraceFS) ListKProbes() (kprobes []Probe, err error) {
+	return dfs.listProbes(kprobeCfgFile)
+}
+
+// ListUProbes lists the currently installed uprobes / uretprobes
+func (dfs *TraceFS) ListUProbes() (uprobes []Probe, err error) {
+	return dfs.listProbes(uprobeCfgFile)
+}
+
+func (dfs *TraceFS) listProbes(filename string) (probes []Probe, err error) {
+	mapping, ok := probeFileInfo[filename]
+	if !ok {
+		return nil, fmt.Errorf("unknown probe events file: %s", filename)
+	}
+	file, err := os.Open(filepath.Join(dfs.basePath, filename))
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		if matches := kprobeRegexp.FindStringSubmatch(scanner.Text()); len(matches) == 6 {
```

Comment:
Should we consider an error if the probe definition doesn't match this regexp?

Reply:
As this is only used to list our own installed kprobes, and there are tests to make sure it detects those, I rather leave it at is so it doesn't break in the event that a different kind of probe is created in the future.